### DISCUSSION TITLE
Acquisition Function to optimize Matheron Path Samples

### DISCUSSION THREAD

**Discussion by jduerholt** [meta-pytorch/botorch#3236] [General]
Hi all,

I have a question regarding Thompson sampling with Matheron path samples. In the example notebook (https://github.com/meta-pytorch/botorch/blob/main/tutorials/thompson_sampling/thompson_sampling.ipynb), the maximizer of the matheron path samples is fined via sampling. But as the matheron path samples are fully differentiable determinisitic models, one should be also able to find their maximizers via `optimize_acqf`. As `optimize_acqf` expects an acquisition function object, which is the most lightweight acqf that you would recommend for this? In the end we just want to apply an objective (for disctinction between maximization/minimization etc). to the path samples and then want to directly optimize on the path sample/deterministic model. I got it running with `qSimpleRegret`, but I have the feeling that there should be a more lightweight solution for this. 

Any ideas?

Best, 

Johannes

**Comment by hvarfner** [comment]
Hi Johannes! Fairly new add, can be found [here](https://github.com/meta-pytorch/botorch/blob/main/botorch/acquisition/thompson_sampling.py). This should have all the functionality that you are looking for, but let us know is something is missing or not working as intended!

**Reply by jduerholt** [reply]
Ah, thank you. Sorry for overlooking this, this is in principle exactly the same as I did with rapping the Matheron model into `qSimpleRegret`, but this implementation is much nicer. Thanks!!!

One other related question: in my use case, I want to generate quite a lot of Matheron samples per a once fitted GP model and then want to optimize it under different constraints, fixed features etc. So with different settings in `optimize_acqf`. This can be of course done with always creating a new `PathwiseThompsonSampling` acqf object, which generates a new Matheron sample. The computationally expensive part in generating Mathron samples is (at least if I understand it correctly) to compute the inverse of `K(X_train, X_train)`  which would then computer over and over again. Do you see an easy possibility to update the code to regenerate the matheron sample for the already instantiated`PathwiseThompsonSampling` acqf object?  If yes, how would you implement it? Ideally one would cache the inverse of `K(X_train, X_train)` and would use it in `redraw`, one could also add a flag if one wants to use the cached inverse or want to recompute ...


Please also correct me, if I am wrong about how the Materon samples are working ;)

**Reply by TobyBoyne** [reply]
Hi @hvarfner, this isn't directly related but I was just reading through the `PathwiseThompsonSampling` class. In the docstring is the following quote:
> Querying the acquisition function gives the summed values of one or more draws from a pathwise drawn posterior sample, and thus it[s] maximization yields one (or multiple) Thompson sample(s).

I'm not sure I understand this correctly. Isn't maximizing the sum of samples equivalent to maximizing the mean of samples, and so you will just be maximizing the *posterior mean* of your model as you let the number of drawn paths increase? And therefore the acqf maximizer won't actually be a Thompson sample? Sorry if I have misunderstood what this acqf does!

**Reply by hvarfner** [reply]
@TobyBoyne It's just me not being sufficiently precise in the docstring. This refers to the sum over `q`, so the acquisition value being returned is the summed values over `q` unique Thompson samples. Description here: https://github.com/meta-pytorch/botorch/blob/main/botorch/acquisition/thompson_sampling.py#L115

**Reply by TobyBoyne** [reply]
Ah I see, thanks for the clarification!

**Reply by hvarfner** [reply]
@jduerholt NP! 

Yes, the expensive part is the inverse of `K(X_train, X_train)`, but you really must have a lot of data for that to be the practical bottleneck, rather than `optimize_acqf`?

With that said, I think caching is doable but not very straightforward. You could imagine writing a cached version of gaussian_update ([this method](https://github.com/meta-pytorch/botorch/blob/main/botorch/sampling/pathwise/update_strategies.py#L91)). I am not sure how it would make its way to the acqf, but I'd be happy to discuss this further.

**Reply by jduerholt** [reply]
Thanks, for the info, for my first prototyping of the algorithm, I will go with the current no-cache version, and if necessary, I will have a look into implementing caching into `gaussian_update`. I will keep you posted!

**Reply by Balandat** [reply]
One thing to keep in mind is that the summing is really because of the limitation of the optimization implementation, see [this comment](https://github.com/meta-pytorch/botorch/blob/main/botorch/acquisition/thompson_sampling.py#L115-L119): 

      # NOTE: The current implementation sums over the q-batch dimension, which means
      # that we are optimizing the sum of independent Thompson samples. In the future,
      # we can leverage *batched* L-BFGS optimization, rather than summing over the q
      # dimension, which will guarantee descent steps for all members of the batch
      # through batch-member-specific learning rate selection.

Let me see if there is a gap or whether we can implement this now.
